ecmwf processing and skill

Tinkaa opened the pull request:
-  Retrieving ECMWF's seasonal forecast with monthly precipitation (`ecmwf_seasonal.py` and `get_ecmwf_seasonal_data.py`)
- Convert the dataset to have all forecasted values for a specific month from different lead times (`read_in_data.py`)
- Get started on skill analysis of MWI (`mwi_ecmwf_monthly_skill.md`)
- Compute values per admin1 instead of raster cell (`read_in_data.py`)

@turnerm I tried to follow "your" GloFas structure as much as possible. Would you be willing to have a look if this code is decent?

turnerm commented on src/indicators/drought/ecmwf_seasonal.py:
```diff
@@ -0,0 +1,301 @@
+
+logger = logging.getLogger(__name__)
+
+class EcmwfSeasonal:
```
There is a lot of repetition between this class and the base GloFAS class -- some of the methods are identical. I think you could probably create a more general base class (maybe a CDS class?), which could go in utils_general or something, and then in flooding / drought we could have the respective glofas / ecmwf subclasses

Tinkaa replied:
Yes agreed! Not having the time for it now, but made an issue #83

turnerm commented on src/indicators/drought/ecmwf_seasonal.py:
```diff
@@ -0,0 +1,301 @@
+        country_name: str,
+        country_iso3: str,
+        area: Area,
+        # leadtimes: List[int],
```
Remove if not used

turnerm commented on src/indicators/drought/ecmwf_seasonal.py:
```diff
@@ -0,0 +1,301 @@
+        self,
+        country_name: str,
+        country_iso3: str,
+        # leadtimes: List[int],
```
Remove if not used

turnerm commented on src/indicators/drought/ecmwf_seasonal.py:
```diff
@@ -0,0 +1,301 @@
+            version=version,
+        )
+
+#TODO: understand if also the case for ecmwf (copied from glofas)
```
You can test this by just calling the xarray expand dims and see if it crashes

Tinkaa replied:
seems to be working fine for me, so will remove. Didn't dig into this, but since you say it should have been fixed, does it have to do anything with versioning that we should indicate in the requirements?

turnerm commented on src/indicators/drought/ecmwf_seasonal.py:
```diff
@@ -0,0 +1,301 @@
+            "month": [str(x).zfill(2) for x in range(1, 13)]
+            if month is None
+            else str(month).zfill(2),
+            "leadtime_month":[str(x) for x in range(1,7)]
```
why 7?

Tinkaa replied:
Good point. This are the leadtimes, now set it as a global var

turnerm commented on src/indicators/drought/ecmwf_seasonal.py:
```diff
@@ -0,0 +1,301 @@
+        """
+
+        def _preprocess_monthly_mean_dataset(ds):
+            dsnew = ds.copy()
```
Why do you need to make a copy?

turnerm commented on src/indicators/drought/ecmwf_seasonal.py:
```diff
@@ -0,0 +1,301 @@
+            #other option could be to convert it to the forecasted time, but gets difficult to concat all different publication dates afterwards
+
+            #todo: make this an input variable!
+            dsnew["step"] = range(1,7)
```
Why 7? If it is a parameter of the dataset then you should get it from the dims, otherwise make it a constant

turnerm commented on src/indicators/drought/ecmwf_seasonal.py:
```diff
@@ -0,0 +1,301 @@
+            #TODO: include way to not make it crash with current year (2021)
+            year_max=2020,
+            cds_name="seasonal-monthly-single-levels",
+            dataset="monthly_mean",
```
The dataset variable expects a list type

turnerm commented on src/malawi/get_ecmwf_seasonal_data.py:
```diff
@@ -0,0 +1,52 @@
+from src.indicators.flooding.glofas.area import AreaFromShape
+from src.indicators.drought.config import Config
+
+# Location of stations on the Jamuna/Brahmaputra river from http://www.ffwc.gov.bd/index.php/googlemap?id=20
```
Remove comment

turnerm commented on src/malawi/get_ecmwf_seasonal_data.py:
```diff
@@ -0,0 +1,52 @@
+# Some lat lon indicated by FFWC are not on the river and have been manually moved to the closest pixel on the river
+# Bahadurabad_glofas corresponds to the control point identified here:
+# https://drive.google.com/file/d/1oNaavhzD2u5nZEGcEjmRn944rsQfBzfz/view
+COUNTRY_NAME = "malawi"
```
Get these from the config file too (I know I didn't, but I should have!)

Tinkaa replied:
You need to define the country name manually I would say, but the country iso3 yes, should come from the config

turnerm commented:
Pausing my review for now -- looks great so far! I think the main thing is that we want to avoid repeating code as much as possible, so I think it's important to consolidate (and possible generalize slightly) any repeated functions in an overarching CDS class. 

I also wrote some tests for the GloFAS class -- you can try to extend those to ECMWF. (Actually I've been trying to improve the tests, I will make a PR and add you as a reviewer!)

On Monday I can look at the notebooks.

turnerm commented on analyses/malawi/notebooks/read_in_data.py:
```diff
@@ -0,0 +1,184 @@
+from src.malawi import get_ecmwf_seasonal_data as gesd
+from src.indicators.drought.config import Config
+
+#TODO: not sure which functions to include here and which in the ecmwf_seasonal file
```
I originally made this file for Bangladesh when I had multiple notebooks using the same methods to avoid duplication. Now that we are using them for other countries, we should probably generalize them somewhere in `src` (although I'm not sure how much overlap there is between GloFAS and ECMWF methods here)

Tinkaa replied:
Moved it to `src/indicators/drought/ecmwf_seasonal/processing.py`. Do you think that makes sense?

turnerm commented on analyses/malawi/notebooks/read_in_data.py:
```diff
@@ -0,0 +1,184 @@
+                ds_date = ds.sel(time=date,number=number,leadtime=leadtime)
+
+                df[["mean_cell", "max_cell", "min_cell"]] = pd.DataFrame(
+                    zonal_stats(vectors=df, raster=ds_date.values, affine=raster_transform, nodata=np.nan))[
```
If speed is an issue, I found it can be faster to use rasterio mask, see this issue [here](https://github.com/OCHA-DAP/pa-COVID-model-parameterization/issues/35)

Tinkaa replied:
Wauw this is great if that speeds up things! Since it is an xarray dataset instead of rasterio it needs some changes, but I think it should be possible. Not the time to look into it now, but will make an issue for it! (also in other scripts we are using `zonal_stats` quite a bit so definitely worth looking into)

Tinkaa commented on src/indicators/drought/ecmwf_seasonal.py:
```diff
@@ -207,26 +197,38 @@ def __init__(self):
+                    logger.info(f"Cannot download data for {year}, because it is in the future")
+            else:
+                month_range = [None]
 
```
@turnerm do you think this is a neat enough way to account for not all months being available for the current year or would you know a programmatically better way?

Tinkaa commented on analyses/malawi/notebooks/mwi_ecmwf_monthly_skill.md:
````diff
@@ -0,0 +1,382 @@
+
+```python
+#interpolate forecast data such that it has the same resolution as the observed values
+#not sure if nearest or linear is most suitable here..
````
Yes I think the forecasts are designed such that edges should be preserved, so will use the nearest

Tinkaa commented:
@turnerm thanks so much for the elaborate review. 
I addressed all the comments and made issues for open questions. 
Can you have another check if it looks decent now? 
The `analyses/malawi/notebooks/mwi_ecmwf_monthly_skill.md` became a lot more elaborate. But only have a look at it if time allows (the `set threshold based on x% probability` is the most relevant section)

turnerm commented on src/indicators/drought/ecmwf_seasonal/ecmwf_seasonal.py:
```diff
@@ -0,0 +1,272 @@
+        self,
+        country_iso3: str,
+        area: Area,
+        leadtimes: List[int] = DEFAULT_LEADTIMES,
```
Default arguments should never be mutable, because if you change the variable somewhere in the method then the default will also be changed. Best practice is to set it to `None`, and then in the method do:
```python
if leadtimes is None:
    leadtimes = DEFAULT_LEADTIMES
```

turnerm commented on analyses/malawi/notebooks/mwi_ecmwf_monthly_skill.md:
```diff
@@ -0,0 +1,946 @@
+import numpy as np
+
+import seaborn as sns
+from mlxtend.evaluate import confusion_matrix
```
This import should be added to the requirements

turnerm commented on analyses/malawi/notebooks/mwi_ecmwf_monthly_skill.md:
```diff
@@ -0,0 +1,946 @@
+    leadtime=leadtime.values).dropna(dim='time')
+    observations = da_obs.reindex({'time': forecast.time}).precip
+    # For all dates
+    crps = xs.crps_ensemble(observations, forecast,member_dim='number')
```
Since we're calculating the CRPS for many different countries now, I've added a method for calculating it [here](https://github.com/OCHA-DAP/pa-anticipatory-action/pull/82/files?file-filters%5B%5D=.md&file-filters%5B%5D=.py&file-filters%5B%5D=.yml#). This script should probably eventually be updated to use it.

Tinkaa replied:
This is really nice! Yes definitely. I am gonna leave it like this for now if that is alright and then get back to it later.

turnerm commented on src/indicators/drought/ecmwf_seasonal/processing.py:
```diff
@@ -0,0 +1,210 @@
+sys.path.append(path_mod)
+
+from src.indicators.drought.ecmwf_seasonal import ecmwf_seasonal
+from src.malawi import get_ecmwf_seasonal_data as gesd
```
To generalize these methods, you shouldn't use Malawi specific information

Tinkaa commented:
Discussed with @turnerm that `src` code is ready to be merged (and opened issues on next steps). `mwi_ecmwf_monthly_skill_dryspells.md` will be reviewed later in a separate branch
